### PR TITLE
core: Add nextgen core routines

### PR DESCRIPTION
**Pull request by Makman2**
The nextgen-core is a scheduler-based system. Bears can offload tasks
via `generate_tasks()` which get executed by a Python pool (in this
case a multiprocess pool).

A dependency mechanism is not implemented yet.

See https://github.com/coala/cEPs/blob/master/cEP-0002.md

### CODE REVIEW

File: coalib/core/Core.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    # FIXME   (for example Mesos).
+
+    # Set up event loop and executor.
+    event_loop = asyncio.SelectorEventLoop()
+    executor = concurrent.futures.ProcessPoolExecutor(
+        max_workers=get_cpu_count())
+
+    # Let's go.
+    schedule_bears(bears, result_callback, event_loop, {}, executor)
+    try:
```
Comment (Nosferatul) [inline coalib/core/Core.py]: Maybe we should have an `except` here too?

Comment (Makman2) [inline coalib/core/Core.py]: No if something crashes here it should bubble up. Exceptions here are likely core related and not bear related and shall not be swallowed

File: coalib/core/Core.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
+import asyncio
+import concurrent.futures
+import functools
+import logging
+import multiprocessing
+
+
+def get_cpu_count():
```
Comment (sils) [inline coalib/core/Core.py]: if this method is only used here, can you mark it with an underscore?

Comment (Makman2) [inline coalib/core/Core.py]: hm it's still functionality someone can use, imo it doesn't have to be "private"

Comment (Makman2) [inline coalib/core/Core.py]: nvm will do so :+1:

File: coalib/core/Core.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    event_loop = asyncio.SelectorEventLoop()
+    executor = concurrent.futures.ProcessPoolExecutor(
+        max_workers=_get_cpu_count())
+
+    # Let's go.
+    schedule_bears(bears, result_callback, event_loop, {}, executor)
+    try:
+        event_loop.run_forever()
+    finally:
+        event_loop.close()
```
Comment (Udayan12167) [inline coalib/core/Core.py]: If we catch all the exceptions properly we can show results for other bears even if one exits out. Is this needed in that case?

Comment (Makman2) [inline coalib/core/Core.py]: If a bear exits or raises an error, other bears still continue to run. Happens here: https://github.com/coala/coala/pull/3871/commits/bb19717550d61898d2c6703354c5ca32c10fe5ba#diff-94f6470cc4dcfa1e33ff5eb050fd5191R112

File: tests/core/CoreTest.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+        run({bear}, on_result)
+
+        on_result.assert_has_calls([unittest.mock.call(i) for i in range(10)],
+                                   any_order=True)
+
+    def test_run_bear_exception(self):
+        # Test exception in bear. Core needs to shutdown directly and not wait
+        # forever.
+        self.execute_run({FailingBear(self.section1, self.filedict1)})
```
Comment (Udayan12167) [inline tests/core/CoreTest.py]: Check exception handling as well?

File: coalib/core/Core.py (hunk 3 above)
Comment (Udayan12167) [inline coalib/core/Core.py]: 👍

File: tests/core/CoreTest.py (hunk 4 above)
Comment (Makman2) [inline tests/core/CoreTest.py]: you mean I shall assert for the log message the core emits?

Comment (Makman2) [inline tests/core/CoreTest.py]: maybe I should also run another bear in another test case and see whether it gets executed properly although the other one crashed^^

Comment (Udayan12167) [inline tests/core/CoreTest.py]: 👍

**Comment by Udayan12167** [comment]
ack https://github.com/coala/coala/pull/3871/commits/ada14dcc9cfd482ac48efe6cc61b9abcb42ed5fb

**Comment by Udayan12167** [comment]
@rultor merge

**Pull request merged**